pkg/auth/csp/selfmanaged.go

```diff
@@ -0,0 +1,200 @@
+	}
+	return nil
+}
+
+func getAuthCodeURL() (string, error) {
+	stateVal, err := state.Generate()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to generate state parameter")
+	}
+	opts := []oauth2.AuthCodeOption{
+		pkceCodePair.Challenge(),
+		pkceCodePair.Method(),
+	}
+
+	return sharedOauthConfig.AuthCodeURL(stateVal.String(), opts...), nil
+}
+
+func GetAccessTokenFromSelfManagedIDP(refreshToken, issuerURL string) (*Token, error) {
+	var mutex sync.Mutex
+	mutex.Lock()
+	defer mutex.Unlock()
+	sharedOauthConfig = &oauth2.Config{
+		RedirectURL:  redirectURL,
+		ClientID:     pinnipedCLIClientID,
+		ClientSecret: "",
+		Scopes:       []string{"openid", "offline_access", "username", "groups"},
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  fmt.Sprintf("%s/%s", issuerURL, authorizationEndpointSuffix),
+			TokenURL: fmt.Sprintf("%s/%s", issuerURL, tokenEndpointSuffix),
+		},
+	}
+
+	if refreshToken != "" {
+		refreshedToken, err := sharedOauthConfig.TokenSource(context.TODO(), &oauth2.Token{RefreshToken: refreshToken}).Token()
+		if err == nil {
+			return &Token{
+				IDToken:      refreshedToken.Extra("id_token").(string),
+				AccessToken:  refreshedToken.AccessToken,
+				RefreshToken: refreshedToken.RefreshToken,
+				// time.Duration is a typedef for int64 and is in nanoseconds
+				// nano seconds till the token expires
+				ExpiresIn: int64(time.Until(refreshedToken.Expiry)),
+				Scope:     loginScopes,
+				TokenType: "id_token",
+			}, nil
+		}
+		log.Debugf("failed to refresh token, err %v", err)
+		// proceed with login flow through the browser
+	}
+	// set the issuer package variable to be used in the callback handler
+	if _, err := url.Parse(issuerURL); err != nil {
+		return nil, errors.Errorf("Issuer URL [%s] is not a valid URL", issuerURL)
+	}
+	var err error
+	if pkceCodePair, err = pkce.Generate(); err != nil {
+		return nil, errors.Wrapf(err, "failed to generate pkce code pair generator")
+	}
+	// perform a browser based login flow if no refreshToken was supplied
+	// or if token refresh failed.
+	g := &errgroup.Group{}
+	g.Go(
+		runLocalListener,
+	)
+	authCodeURL, err := getAuthCodeURL()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to generate authcode url for OIDC provider at %s, err= %v", sharedOauthConfig.Endpoint.AuthURL, err)
+	}
+
+	fmt.Printf("Please open this URL in a browser window to complete the login\n\t %s\n", authCodeURL)
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	if token == nil || token.Extra("id_token").(string) == "" {
+		return nil, errors.Errorf("token issuer %s did not return expected tokens", issuerURL)
+	}
+
+	return &Token{
+		IDToken:      token.Extra("id_token").(string),
+		AccessToken:  token.AccessToken,
+		RefreshToken: token.RefreshToken,
+		// time.Duration is a typedef for int64 and is in nanoseconds
+		// nano seconds till the token expires
+		ExpiresIn: int64(time.Until(token.Expiry)),
```

Comment:
I think we should convert `token.Expiry` to seconds, because `ExpiresIn` should be in seconds according to its doc and is used as seconds everywhere else in the code.

Reply:
Updated. Thanks